### PR TITLE
Afbase/tcp connect

### PR DESCRIPTION
**Pull request by afbase**
briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this
  change?)
 - Adding a very simple tcpconnect example
* what changes does this pull request make?
 - src/tcpconnect folder contents + new bin target in Cargo.toml
* are there any non-obvious implications of these changes? (does it break compatibility with previous
  versions, etc)
```
warning: trait objects without an explicit `dyn` are deprecated
  --> src/tcpconnect/main.rs:59:39
   |
59 | fn perf_ipv4_data_t_callback() -> Box<FnMut(&[u8]) + Send> {
   |                                       ^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn FnMut(&[u8]) + Send`
   |
   = note: `#[warn(bare_trait_objects)]` on by default

warning: trait objects without an explicit `dyn` are deprecated
  --> src/tcpconnect/main.rs:78:39
   |
78 | fn perf_ipv6_data_t_callback() -> Box<FnMut(&[u8]) + Send> {
   |                                       ^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn FnMut(&[u8]) + Send`

```
I get a warning about my callback functions suggesting I use `dyn`.  I'm still trying to understand this issue.

I tested on ubuntu 18.04 for both ipv6 and ipv4 addresses.

### CODE REVIEW

**Comment by brayniac** [comment]
@afbase - just FYI - working to get CI setup so PRs get tested before merging. I'll update here once that's landed.

**Comment by brayniac** [comment]
@afbase - #2 has landed. You'll want to merge into your feature branch. I'd like to see running this new binary in CI jobs as that provides us some good matrix coverage across bcc versions and os/compiler versions. 

As for the warning you got about `dyn` you need to prefix the `FnMut` to make it `dyn FnMut` on those lines you're getting warnings for.

File: src/tcpconnect/main.rs [hunk 1]
```diff
@@ -0,0 +1,270 @@
+use bcc::core::BPF;
```
Comment (brayniac) [inline src/tcpconnect/main.rs:1]: Please run rustfmt.

File: src/tcpconnect/main.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+use std::net::{Ipv4Addr, Ipv6Addr};
+use clap::{App, Arg};
+use failure::Error;
+use std::process;
+use core::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::{ptr, str};
+extern crate ctrlc;
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: I don't believe this is needed.

File: src/tcpconnect/tcpconnect.c [hunk 3]
```diff
@@ -0,0 +1,108 @@
+#include <uapi/linux/ptrace.h>
```
Comment (brayniac) [inline src/tcpconnect/tcpconnect.c]: I would rename this file to `bpf.c`, naming it after the tool is redundant considering it's in its own folder in this project.

File: src/tcpconnect/tcpconnect.c [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+int trace_connect_v4_return(struct pt_regs *ctx)
+{
+    return trace_connect_return(ctx, 4);
+}
+
+int trace_connect_v6_return(struct pt_regs *ctx)
+{
+    return trace_connect_return(ctx, 6);
+}
```
Comment (brayniac) [inline src/tcpconnect/tcpconnect.c]: A trailing newline should be added

File: src/tcpconnect/main.rs [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+const DEBUG: bool = false;
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: I think we should get rid of this. I'm not sure that we need to provide the functionality of printing the bpf code, but I definitely don't like this as a const.

File: src/tcpconnect/main.rs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                let uuid_set = format!("if (uid != {}) {{ return 0; }}",uuid);
+                bpf_text.replace("FILTER_UID", uuid_set.as_str())
+            };
+        }
+        // replace remaining unset template flags, if any
+        bpf_text = bpf_text.replace("FILTER_PID", "");
+        bpf_text = bpf_text.replace("FILTER_PORT", "");
+        bpf_text = bpf_text.replace("FILTER_UID", "");
+
+        let ebpf = matches.is_present("ebpf");
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: I think we can probably get rid of this

File: src/tcpconnect/main.rs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        .arg(
+            Arg::with_name("p")
+            .long("pid")
+            .help("trace this PID only")
+            .value_name("PID")
+            .number_of_values(1)
+            .takes_value(true)
+        )
+        .arg(
+            Arg::with_name("ebpf")
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: This can be dropped if we don't need the debug functionality.

File: src/tcpconnect/main.rs [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+            // trace pid
+            let pid = p.parse::<u32>().expect("pid should be a non-negative integer");
+            bpf_text = {
+                let uuids = format!("if (pid != {}) {{ return 0; }}", pid);
+                bpf_text.replace("FILTER_PID", uuids.as_str())
+            };
+        }
+        // handle ports 
+        let port_listings = matches.values_of("P");
+        if port_listings.is_some() {
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: change to `if let Some(port_listings) = matches.values_of("P") {`

File: src/tcpconnect/main.rs [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        if let Some(p) = matches.value_of("p") {
+            // trace pid
+            let pid = p.parse::<u32>().expect("pid should be a non-negative integer");
+            bpf_text = {
+                let uuids = format!("if (pid != {}) {{ return 0; }}", pid);
+                bpf_text.replace("FILTER_PID", uuids.as_str())
+            };
+        }
+        // handle ports 
+        let port_listings = matches.values_of("P");
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: I don't think we need this

File: src/tcpconnect/main.rs [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        if port_listings.is_some() {
+            let mut port_listings_raw = port_listings.expect("comma separated listing of ports, e.g. 1,2,3,80");
+            let mut ports_conditions: Vec<String> = Vec::new();
+            loop{
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: `for port in port_listings {`

You don't need to loop and match

File: src/tcpconnect/main.rs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+                match port_listings_raw.next() {
+                    Some(port) => {
+                        let port_u16 = port.parse::<u16>().expect("port number between 1-65535");
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: if we allow a comma separated listing of ports, does this parsing work as expected?

**Review by brayniac — CHANGES_REQUESTED** [review]

File: src/tcpconnect/main.rs (hunk 11 above)
Comment (afbase) [reply src/tcpconnect/main.rs]: image

so it will support comma separated.  I changed my description anyway.

File: src/tcpconnect/main.rs (hunk 6 above)
Comment (afbase) [reply src/tcpconnect/main.rs]: I got rid of the epbf flag since we can remove debug functionality.

File: src/tcpconnect/main.rs (hunk 5 above)
Comment (afbase) [reply src/tcpconnect/main.rs]: I got rid of DEBUG

File: src/tcpconnect/main.rs (hunk 2 above)
Comment (afbase) [reply src/tcpconnect/main.rs]: I followed some of the other examples. What would be problematic about using these?

File: src/tcpconnect/main.rs (hunk 1 above)
Comment (afbase) [reply src/tcpconnect/main.rs:1]: TIL `rustfmt`

File: src/tcpconnect/main.rs (hunk 2 above)
Comment (brayniac) [reply src/tcpconnect/main.rs]: It's the `extern crate ctrlc;` that I think can be removed. We're not using any macros from it, and extern crate is typically not needed in Rust 2018 edition.

File: src/tcpconnect/main.rs [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    // handle ports
+    if let Some(port_listings) = matches.values_of("P") {
+        let mut ports_conditions: Vec<String> = Vec::new();
+        for port in port_listings {
+            let port_u16 = port.parse::<u16>().expect("port number between 1-65535");
+            let port_ntohs = ntohs(port_u16);
+            let port_condition = format!("dport != {}", port_ntohs);
+            ports_conditions.push(port_condition);
+        }
+        let port_contition_statement = ports_conditions.join(" && ");
```
Comment (brayniac) [inline src/tcpconnect/main.rs]: What's up with the spelling for this? You could probably move the `join()` down into the `format!` below

File: .travis.yml [hunk 13]
```diff
@@ -0,0 +1,864 @@
...
+        - cd _build
+        - cmake .. -DCMAKE_INSTALL_PREFIX=/usr
+        - make
+        - sudo make install
+        - cd ../..
+        - cargo build --features $BCC
+        - cargo test --features $BCC
+        - cargo build --release --features $BCC
+        - cargo test --release --features $BCC
+        - sudo target/release/runqlat --interval 1 --windows 5
```
Comment (brayniac) [inline .travis.yml]: You need to add your binary here as well. Will require wrapping with `timeout` command. See github.com/twitter/rezolus for example. 

I'm unsure why the diff for this file is so large, since we do have the CI config in master now.

Comment (afbase) [reply .travis.yml]: @brayniac 
With the `timeout` command, is there a time limit we wish to set for this.  I've placed a 1 minute limit for this to run for now.  I wonder how this will impact the travis CI free build time budget per month.

Comment (brayniac) [reply .travis.yml]: @afbase - I am only aware of limits per-job. I believe it's important we're only releasing tested code. From the docs, I can only see limits per-job, where the build will not timeout unless any job within it does. (https://docs.travis-ci.com/user/customizing-the-build#build-timeouts)

If we look for the new tool, we've caught an issue on bcc 0.4.0: https://travis-ci.org/rust-bpf/bpf-tools/jobs/656075826#L1857

It's possible we'll want to restrict our compatibility to some subset of recent bcc versions, but illustrates the importance of having a ci matrix and running our tools through it.

Comment (afbase) [reply .travis.yml]: > illustrates the importance of having a ci matrix and running our tools through it.

1. I really like the build matrix you have going on here. 

>we'll want to restrict our compatibility to some subset of recent bcc versions

would we have to restrict the build for tcpconnect to exclude `bcc_v0_4_0`?  Could you elaborate a bit more on this?  

>If we look for the new tool, we've caught an issue on bcc 0.4.0: https://travis-ci.org/rust-bpf/bpf-tools/jobs/656075826#L1857

1. TIL travis gives line numbers for build STDOUT!  this is a super handy feature.

Comment (brayniac) [reply .travis.yml]: I was trying to say that we might want to drop anything older than maybe bcc 0.10.0 for the whole repo. If we can make the tcpconnect tool work on bcc 0.4.0, that would be preferred until we decide on a particular oldest bcc version that we want to support the tools on.

Comment (afbase) [reply .travis.yml]: I'm still playing with `bpf.c` to see if I can get a variant to work with bcc 0.4.0. Giving myself a little more time.

Comment (afbase) [reply .travis.yml]: hi @brayniac,
> was trying to say that we might want to drop anything older than maybe bcc 0.10.0 for the whole repo. If we can make the tcpconnect tool work on bcc 0.4.0

I'm leaning towards dropping support for anything older than 0.10.0.  I've spent a considerable amount of time on trying to get bcc 0.4.0 working but no such luck and I'd like to spend time on another project for ~1 month (I'll come back to bpf-tools and add more thereafter).

The code seems to work as low as bcc 0.5.0 for the time being.  

My thoughts are as follows:
1. remove tests in `travis.yml` for any `bcc 0.x.0` for some version less than or equal to `x`. 
1. version control bcc in the Cargo.toml dependencies. 

Let me know what you think.

Comment (brayniac) [reply .travis.yml]: I think it's fair to drop anything pre 0.10.0. Should be able to simply remove prior versions from travis config and from the Cargo.toml features.

**Comment by brayniac** [comment]
@afbase - this might be easier if I open a PR and merge some improvements to the CI - I'm going to follow the pattern used in github.com/twitter/rezolus - for that project, we decided to support 0.8.0+ - as this gets support on centos 7 and debian stable too - which I think is beneficial for a tools project as well. Let me open a PR for that, you'll need to rebase - but hopefully that helps limit the scope of your change here

**Comment by brayniac** [comment]
@afbase #3 is merged

**Comment by afbase** [comment]
Bad News: Covid-19 has curtailed my other projects,
Good News: I'm back onto this one!  I'm thinking to work on more TCP stuff + cgroups in new PRs.

Let me know if there are other changes to make for this PR.

**Comment by brayniac** [comment]
> Bad News: Covid-19 has curtailed my other projects,
> Good News: I'm back onto this one! I'm thinking to work on more TCP stuff + cgroups in new PRs.
> 
> Let me know if there are other changes to make for this PR.

There's still something weird with this diff - I think there should only be new files, adding the bin in the cargo manifest, and a couple additional lines in the `build/ci.sh` file. It may be easier to recreate the diff from master and force-push over your current feature branch.

File: .travis.yml [hunk 14]
```diff
@@ -0,0 +1,207 @@
+language: rust
```
Comment (brayniac) [inline .travis.yml]: There should be no changes to this file in your diff.

File: Cargo.toml [hunk 15]
```diff
@@ -20,15 +20,15 @@ failure = "0.1.5"
 name = "runqlat"
 path = "src/runqlat/main.rs"
 
+[[bin]]
+name = "tcpconnect"
+path = "src/tcpconnect/main.rs"
+
 [features]
 static = ["bcc/static"]
-bcc_v0_4_0 = ["bcc/v0_4_0"]
```
Comment (brayniac) [inline Cargo.toml]: Not sure why the changes starting from here are showing in this diff. But I think should only be adding the binary target above.

File: build/ci.sh [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+fi
+
+## Build and test
+## Build and test
+if [ -n "${FEATURES}" ]; then
+    cargo build --features "${FEATURES}"
+    cargo test --features "${FEATURES}"
+    cargo build --release --features "${FEATURES}"
+    cargo test --release --features "${FEATURES}"
+    sudo target/release/runqlat --interval 1 --windows 5
```
Comment (brayniac) [inline build/ci.sh:64]: Need to add running the binary here

File: build/ci.sh [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+else
+    cargo build
+    cargo test
+    cargo build --release
+    cargo test --release
+    sudo target/release/runqlat --interval 1 --windows 5
```
Comment (brayniac) [inline build/ci.sh:72]: And here. But there should be no other changes to this file

File: src/runqlat/main.rs [hunk 18]
```diff
@@ -10,31 +10,6 @@ use std::{mem, thread, time};
 //
 // Based on: https://github.com/iovisor/bcc/blob/master/tools/runqlat.py
 
-#[cfg(any(
```
Comment (brayniac) [inline src/runqlat/main.rs]: I don't think there should be any change to this file?

**Review by brayniac — CHANGES_REQUESTED** [review]

**Review by brayniac — COMMENTED** [review]
Thanks for sticking with this! I'll review more thoroughly tomorrow. The diff itself looks good now.

**Comment by afbase** [comment]
thanks for letting me participate and for your time!

**Review by brayniac — APPROVED** [review]
